### ISSUE TITLE
Fragment spreads are possible rule exception

### ISSUE THREAD

**Issue by dinsley** [rmosolgo/graphql-ruby#583]
I've been attempting to use fragments with `graphql-client` and am getting the attached backtrace when validation takes place.

The `schema` that I'm linking too has been generated by [Absinthe](https://github.com/absinthe-graphql/absinthe).
 
The `schema` and code used to replicate the issue is available here:

https://gist.github.com/dinsley/4546061a94f1ba2e750599fdda121a63

There was some discussion around this on the GraphQL Slack, it started here:

https://graphql.slack.com/archives/ruby/p1488915326002318

## Backtrace

```
/usr/local/bundle/gems/graphql-1.4.5/lib/graphql/static_validation/rules/fragment_spreads_are_possible.rb:26:in `block (2 levels) in validate': undefined method `type' for nil:NilClass (NoMethodError)
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/static_validation/rules/fragment_spreads_are_possible.rb:25:in `each'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/static_validation/rules/fragment_spreads_are_possible.rb:25:in `block in validate'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:65:in `block in apply_hooks'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:65:in `each'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:65:in `reduce'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:65:in `apply_hooks'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:60:in `end_visit'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:49:in `visit_node'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/language/visitor.rb:39:in `visit'
	from /usr/local/bundle/gems/graphql-1.4.5/lib/graphql/static_validation/validator.rb:34:in `validate'
	from /usr/local/bundle/gems/graphql-client-0.6.1/lib/graphql/client.rb:240:in `parse'
	from /iqapp/lib/integrations/cerebro/proxy/graphql.rb:28:in `<module:Graphql>'
	from /iqapp/lib/integrations/cerebro/proxy/graphql.rb:8:in `<module:Cerebro>'
	from /iqapp/lib/integrations/cerebro/proxy/graphql.rb:7:in `<module:Integrations>'
	from /iqapp/lib/integrations/cerebro/proxy/graphql.rb:6:in `<top (required)>'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `require'
	from /iqapp/lib/integrations/cerebro/adapter.rb:4:in `<top (required)>'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `require'
	from /iqapp/config/initializers/integrations.rb:27:in `<top (required)>'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:268:in `load'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:268:in `block in load'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:268:in `load'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/engine.rb:652:in `block in load_config_initializer'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/notifications.rb:166:in `instrument'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/engine.rb:651:in `load_config_initializer'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/engine.rb:616:in `block (2 levels) in <class:Engine>'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/engine.rb:615:in `each'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/engine.rb:615:in `block in <class:Engine>'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/initializable.rb:30:in `instance_exec'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/initializable.rb:30:in `run'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/initializable.rb:55:in `block in run_initializers'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:228:in `block in tsort_each'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:350:in `block (2 levels) in each_strongly_connected_component'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:422:in `block (2 levels) in each_strongly_connected_component_from'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:431:in `each_strongly_connected_component_from'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:421:in `block in each_strongly_connected_component_from'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/initializable.rb:44:in `each'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/initializable.rb:44:in `tsort_each_child'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:415:in `call'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:415:in `each_strongly_connected_component_from'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:349:in `block in each_strongly_connected_component'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:347:in `each'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:347:in `call'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:347:in `each_strongly_connected_component'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:226:in `tsort_each'
	from /usr/local/lib/ruby/2.3.0/tsort.rb:205:in `tsort_each'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/initializable.rb:54:in `run_initializers'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/application.rb:352:in `initialize!'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/railtie.rb:194:in `public_send'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/railtie.rb:194:in `method_missing'
	from /iqapp/config/environment.rb:6:in `<top (required)>'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `block in require'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:240:in `load_dependency'
	from /usr/local/bundle/gems/activesupport-4.2.7.1/lib/active_support/dependencies.rb:274:in `require'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/application.rb:328:in `require_environment!'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:142:in `require_application_and_environment!'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:67:in `console'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/commands/commands_tasks.rb:39:in `run_command!'
	from /usr/local/bundle/gems/railties-4.2.7.1/lib/rails/commands.rb:17:in `<top (required)>'
	from script/rails:7:in `require'
	from script/rails:7:in `<main>'
````

**Comment by rmosolgo** [comment]
Thanks for providing all these details! 

Ive written some test cases: 

- 1.4.x branch: https://github.com/rmosolgo/graphql-ruby/compare/1.4.x...client-test-1.4.5
- master: https://github.com/rmosolgo/graphql-ruby/compare/client-test

It's a bit weird because 1.4.x has failing tests where things were back-ported, so I'm still trying to work it out, but I've got a replicated failure locally!

**Comment by rmosolgo** [comment]
For me, it fails on 1.4.x but passes on master. Can you try pointing at github in your gemfile?

```ruby 
gem "graphql", github: "rmosolgo/graphql-ruby"
```

I think this change fixed it: https://github.com/rmosolgo/graphql-ruby/commit/41157ab8e18e6e7c54daad88d33f31d3e39a3471#diff-6ba53817f2a690cec2a85730d1f72cdeL26

Previously, methods were called _before_ checking the looked-up value, but now, we only proceed with method calls when the lookup returns a truthy value: 

```diff
- fragment_child_name = context.fragments[frag_spread.node.name].type.name
+ frag_node = context.fragments[frag_spread.node.name]
+ if frag_node
+   fragment_child_name = frag_node.type.name
```

**Comment by rmosolgo** [comment]
I expect this to be fixed in 1.5.0, please reopen if you see this bug in that version or later!
